plugins/inputs/diskio/diskio.go

```diff
@@ -127,16 +127,32 @@ func (d *DiskIO) Gather(acc telegraf.Accumulator) error {
 			"merged_writes":    io.MergedWriteCount,
 		}
 		if lastValue, exists := d.lastIOCounterStat[k]; exists {
-			deltaRWCount := float64(io.ReadCount + io.WriteCount - lastValue.ReadCount - lastValue.WriteCount)
-			deltaRWTime := float64(io.ReadTime + io.WriteTime - lastValue.ReadTime - lastValue.WriteTime)
-			deltaIOTime := float64(io.IoTime - lastValue.IoTime)
-			if deltaRWCount > 0 {
-				fields["io_await"] = deltaRWTime / deltaRWCount
-				fields["io_svctm"] = deltaIOTime / deltaRWCount
-			}
-			itv := float64(collectTime.Sub(d.lastCollectTime).Milliseconds())
-			if itv > 0 {
-				fields["io_util"] = 100 * deltaIOTime / itv
+			// Check for counter wraparound before calculating deltas
+			rwWrapped := isCounterWraparound(io.ReadCount+io.WriteCount, lastValue.ReadCount+lastValue.WriteCount)
+			timeWrapped := isCounterWraparound(io.ReadTime+io.WriteTime, lastValue.ReadTime+lastValue.WriteTime)
+			ioTimeWrapped := isCounterWraparound(io.IoTime, lastValue.IoTime)
+
+			if rwWrapped || timeWrapped || ioTimeWrapped {
+				// Provide zero values instead of omitting fields to avoid sparse metrics
+				fields["io_await"] = 0.0
+				fields["io_svctm"] = 0.0
+				fields["io_util"] = 0.0
+
+				d.Log.Debugf("Counter wraparound detected for %s, providing zero values", io.Name)
+			} else {
+				deltaRWCount := float64(io.ReadCount + io.WriteCount - lastValue.ReadCount - lastValue.WriteCount)
+				deltaRWTime := float64(io.ReadTime + io.WriteTime - lastValue.ReadTime - lastValue.WriteTime)
+				deltaIOTime := float64(io.IoTime - lastValue.IoTime)
+
+				if deltaRWCount > 0 {
+					fields["io_await"] = deltaRWTime / deltaRWCount
+					fields["io_svctm"] = deltaIOTime / deltaRWCount
+				}
+
+				itv := float64(collectTime.Sub(d.lastCollectTime).Milliseconds())
+				if itv > 0 {
+					fields["io_util"] = 100 * deltaIOTime / itv
+				}
 			}
```

Comment:
@skartikey reading the previous code again it seems that the problem is that the `float64` conversion is done _after_ computation instead of converting the individual values first. So how about

```suggestion
			// Check for wrap around
			if io.ReadCount >= lastValue.ReadCount && io.WriteCount >= lastValue.WriteCount {
				deltaRWCount := float64(io.ReadCount - lastValue.ReadCount) + float64(io.WriteCount  - lastValue.WriteCount)
				deltaRWTime := float64(io.ReadTime + io.WriteTime - lastValue.ReadTime - lastValue.WriteTime)
				deltaIOTime := float64(io.IoTime - lastValue.IoTime)
				fields["io_await"] = deltaRWTime / deltaRWCount
				fields["io_svctm"] = deltaIOTime / deltaRWCount
			}
			itv := float64(collectTime.Sub(d.lastCollectTime).Milliseconds())
			if itv > 0 {
				fields["io_util"] = 100 * deltaIOTime / itv
			}
```

Reply:
@srebhan Thanks for the review, I like your approach since it’s simpler, easier to maintain, and follows the proven pattern most monitoring tools use. The direct comparison makes the wrap detection very clear and avoids over-complicating things.

That said, I’m still leaning toward keeping the 0.0 assignment instead of omitting fields, so that we don’t end up with sparse metrics during wrap events.